File: server/src/providerCatalog.ts
type ProviderType = "openai" | "anthropic" | "google";

export type ProviderCatalogEntry = {
  model: string;
  providerType: ProviderType;
  inputUsdPerMillion: string;
  outputUsdPerMillion: string;
  supportsStreaming: boolean;
};

const CATALOG: ProviderCatalogEntry[] = [
  {
    model: "gpt-4.1-mini",
    providerType: "openai",
    inputUsdPerMillion: "0.40",
    outputUsdPerMillion: "1.60",
    supportsStreaming: true,
  },
  {
    model: "gpt-4o-mini",
    providerType: "openai",
    inputUsdPerMillion: "0.15",
    outputUsdPerMillion: "0.60",
    supportsStreaming: true,
  },
  {
    model: "gpt-4o",
    providerType: "openai",
    inputUsdPerMillion: "2.50",
    outputUsdPerMillion: "10.00",
    supportsStreaming: true,
  },
  {
    model: "claude-3-5-sonnet-latest",
    providerType: "anthropic",
    inputUsdPerMillion: "3.00",
    outputUsdPerMillion: "15.00",
    supportsStreaming: true,
  },
  {
    model: "claude-3-5-haiku-latest",
    providerType: "anthropic",
    inputUsdPerMillion: "0.80",
    outputUsdPerMillion: "4.00",
    supportsStreaming: true,
  },
  {
    model: "claude-3-haiku",
    providerType: "anthropic",
    inputUsdPerMillion: "0.25",
    outputUsdPerMillion: "1.25",
    supportsStreaming: true,
  },
  {
    model: "gemini-1.5-pro",
    providerType: "google",
    inputUsdPerMillion: "3.50",
    outputUsdPerMillion: "10.50",
    supportsStreaming: true,
  },
  {
    model: "gemini-1.5-flash",
    providerType: "google",
    inputUsdPerMillion: "0.35",
    outputUsdPerMillion: "0.70",
    supportsStreaming: true,
  },
];

export function defaultCatalogForProvider(type: ProviderType): ProviderCatalogEntry[] {
  return CATALOG.filter((entry) => entry.providerType === type);
}

export function findCatalogEntry(model: string): ProviderCatalogEntry | null {
  const normalized = model.trim().toLowerCase();
  if (normalized === "gpt-fallback-demo") {
    return findCatalogEntry("gpt-4o-mini");
  }
  return (
    CATALOG.find((entry) => entry.model.toLowerCase() === normalized) ?? null
  );
}

export function inferProviderTypeFromModel(model: string): ProviderType | null {
  const normalized = model.trim().toLowerCase();
  if (normalized === "gpt-fallback-demo") return "openai";
  if (normalized.startsWith("gpt-")) return "openai";
  if (normalized.startsWith("claude")) return "anthropic";
  if (normalized.startsWith("gemini")) return "google";
  return null;
}

File: backend-java/src/main/java/com/taas/gateway/ProviderCatalog.java
package com.taas.gateway;

import java.math.BigDecimal;
import java.util.List;

public final class ProviderCatalog {
  private ProviderCatalog() {
  }

  public static final List<Entry> ENTRIES =
      List.of(
          new Entry("gpt-4.1-mini", "openai", new BigDecimal("0.40"), new BigDecimal("1.60"), true),
          new Entry("gpt-4o-mini", "openai", new BigDecimal("0.15"), new BigDecimal("0.60"), true),
          new Entry("gpt-4o", "openai", new BigDecimal("2.50"), new BigDecimal("10.00"), true),
          new Entry("claude-3-5-sonnet-latest", "anthropic", new BigDecimal("3.00"), new BigDecimal("15.00"), true),
          new Entry("claude-3-5-haiku-latest", "anthropic", new BigDecimal("0.80"), new BigDecimal("4.00"), true),
          new Entry("claude-3-haiku", "anthropic", new BigDecimal("0.25"), new BigDecimal("1.25"), true),
          new Entry("gemini-1.5-pro", "google", new BigDecimal("3.50"), new BigDecimal("10.50"), true),
          new Entry("gemini-1.5-flash", "google", new BigDecimal("0.35"), new BigDecimal("0.70"), true));

  public static Entry find(String model) {
    String normalized = normalizeModel(model);
    return ENTRIES.stream().filter(entry -> entry.model().equalsIgnoreCase(normalized)).findFirst().orElse(null);
  }

  public static String inferProviderType(String model) {
    String normalized = normalizeModel(model).toLowerCase();
    if (normalized.startsWith("gpt-")) return "openai";
    if (normalized.startsWith("claude")) return "anthropic";
    if (normalized.startsWith("gemini")) return "google";
    return null;
  }

  private static String normalizeModel(String model) {
    return model;
  }

  public record Entry(
      String model,
      String providerType,
      BigDecimal inputUsdPerMillion,
      BigDecimal outputUsdPerMillion,
      boolean supportsStreaming) {
  }
}
